File: frontend/src/api/sequences.ts
import { api } from "./client";

export type SequenceStatus = "draft" | "active" | "paused" | "archived";
export type StepChannel =
  | "email"
  | "linkedin_dm"
  | "linkedin_connect"
  | "linkedin_like"
  | "call"
  | "sms"
  | "whatsapp";

export type BranchCondition = "replied" | "opened" | "clicked" | "default";
export interface StepTransition {
  on: BranchCondition;
  to_step_id: number | null; // null = stop the sequence
}

export interface StepOut {
  id: number;
  sequence_id: number;
  step_order: number;
  channel: StepChannel;
  delay_days: number;
  delay_hours: number;
  subject: string | null;
  body: string;
  config: Record<string, unknown>;
  transitions?: StepTransition[];
  created_at: string;
  updated_at: string;
}

export interface SequenceOut {
  id: number;
  name: string;
  description: string | null;
  status: SequenceStatus;
  timezone: string;
  send_window_start: string; // "HH:MM:SS"
  send_window_end: string;
  send_days_mask: number;
  ai_followups_enabled: boolean;
  track_opens: boolean;
  track_clicks: boolean;
  ai_knowledge_id?: string | null;
  step_count: number;
  active_enrolments: number;
  created_at: string;
  updated_at: string;
}

export interface SequenceDetail extends SequenceOut {
  steps: StepOut[];
}

export interface SequenceCreate {
  name: string;
  description?: string | null;
  timezone?: string;
  send_window_start?: string;
  send_window_end?: string;
  send_days_mask?: number;
  ai_followups_enabled?: boolean;
  track_opens?: boolean;
  track_clicks?: boolean;
}

export interface StepCreate {
  channel: StepChannel;
  step_order?: number;
  delay_days?: number;
  delay_hours?: number;
  subject?: string | null;
  body: string;
  config?: Record<string, unknown>;
}

export interface EnrolmentResult {
  enrolled: number;
  deduped: number;
  skipped_no_identity: number;
}

export interface SequenceGenerateResponse {
  sequence: SequenceDetail;
  warnings: string[];
}

export type GroundingVerdict = "grounded" | "weak" | "possible_hallucination";

export interface GroundingDocument {
  id: number;
  filename: string;
  status: string;
  indexed: boolean;
}

export interface StepGrounding {
  step_id: number;
  step_order: number;
  channel: StepChannel;
  subject: string | null;
  similarity: number;
  mean_top_k: number;
  supported_ratio: number;
  verdict: GroundingVerdict;
  best_chunk_filename: string | null;
}

export interface SequenceGrounding {
  sequence_id: number;
  documents: GroundingDocument[];
  has_indexed_docs: boolean;
  chunk_count: number;
  avg_similarity: number;
  steps: StepGrounding[];
  note: string | null;
}

export async function getSequenceGrounding(id: number) {
  return (await api.get<SequenceGrounding>(`/sequences/${id}/grounding`)).data;
}

export async function listSequences() {
  return (await api.get<SequenceOut[]>("/sequences")).data;
}
export async function getSequence(id: number) {
  return (await api.get<SequenceDetail>(`/sequences/${id}`)).data;
}
export async function createSequence(dto: SequenceCreate) {
  return (await api.post<SequenceOut>("/sequences", dto)).data;
}
export async function generateSequenceFromPrompt(dto: {
  prompt: string;
  timezone?: string;
  document_ids?: number[];
}) {
  return (await api.post<SequenceGenerateResponse>("/sequences/generate", dto)).data;
}

export interface SequenceGenerateRequest {
  name: string;
  description: string;
  channels: ("email" | "linkedin")[];
  num_emails?: number;
  timezone?: string;
  test_mode?: boolean;
}
export async function generateSequence(dto: SequenceGenerateRequest) {
  return (await api.post<SequenceDetail>("/sequences/generate-quick", dto)).data;
}
export async function updateSequence(id: number, dto: Partial<SequenceCreate>) {
  return (await api.patch<SequenceOut>(`/sequences/${id}`, dto)).data;
}
export async function changeStatus(id: number, status: SequenceStatus) {
  return (await api.post<SequenceOut>(`/sequences/${id}/status`, { status })).data;
}

export interface TestNowResult { fired: number; activated?: boolean; enrolment_ids: number[]; }
export async function testNow(id: number) {
  return (await api.post<TestNowResult>(`/sequences/${id}/test-now`)).data;
}
export async function deleteSequence(id: number) {
  await api.delete(`/sequences/${id}`);
}

export async function addStep(sequenceId: number, dto: StepCreate) {
  return (await api.post<StepOut>(`/sequences/${sequenceId}/steps`, dto)).data;
}
export async function updateStep(sequenceId: number, stepId: number, dto: StepCreate) {
  return (await api.patch<StepOut>(`/sequences/${sequenceId}/steps/${stepId}`, dto)).data;
}
export async function deleteStep(sequenceId: number, stepId: number) {
  await api.delete(`/sequences/${sequenceId}/steps/${stepId}`);
}
export async function reorderSteps(sequenceId: number, stepIds: number[]) {
  return (await api.post<StepOut[]>(`/sequences/${sequenceId}/steps/reorder`, { step_ids: stepIds })).data;
}

// ── A/B testing ──────────────────────────────────────────────────────────────
export interface AbVariant {
  label: string;
  subject?: string | null;
  body: string;
  weight: number;
}
export interface AbVariantStat {
  label: string;
  sent: number;
  opened: number;
  clicked: number;
  replied: number;
  positive: number;
  open_rate: number;
  click_rate: number;
  reply_rate: number;
  positive_rate: number;
}
export interface AbStatsResponse {
  configured: AbVariant[];
  stats: AbVariantStat[];
  winner: string | null;
}
export async function getAbStats(sequenceId: number, stepId: number) {
  return (await api.get<AbStatsResponse>(`/sequences/${sequenceId}/steps/${stepId}/ab-stats`)).data;
}
export async function promoteVariant(sequenceId: number, stepId: number, label: string) {
  return (await api.post<StepOut>(`/sequences/${sequenceId}/steps/${stepId}/ab-promote`, { label })).data;
}

// ── Conditional branching ────────────────────────────────────────────────────
export async function setStepTransitions(sequenceId: number, stepId: number, transitions: StepTransition[]) {
  return (await api.put<StepOut>(`/sequences/${sequenceId}/steps/${stepId}/transitions`, { transitions })).data;
}

export async function enrolLeads(sequenceId: number, leadIds: number[]) {
  return (await api.post<EnrolmentResult>(`/sequences/${sequenceId}/enrol`, { lead_ids: leadIds })).data;
}

File: backend/src/outreach/schemas/sequences.py
from datetime import datetime, time
from typing import Annotated, Literal, Union

from pydantic import BaseModel, Field

# ---------- Steps (discriminated union per channel) ----------


class _StepBase(BaseModel):
    step_order: int | None = Field(default=None, ge=1, le=50)
    delay_days: int = Field(default=0, ge=0, le=365)
    delay_hours: int = Field(default=0, ge=0, le=23)
    config: dict = Field(default_factory=dict)
    transitions: list[dict] = Field(default_factory=list)


class EmailStepCreate(_StepBase):
    channel: Literal["email"]
    subject: str = Field(min_length=1, max_length=250)
    body: str = Field(min_length=1, max_length=16000)


class LinkedInDmStepCreate(_StepBase):
    channel: Literal["linkedin_dm"]
    subject: None = None
    body: str = Field(min_length=1, max_length=8000)


class LinkedInConnectStepCreate(_StepBase):
    channel: Literal["linkedin_connect"]
    subject: None = None
    body: str = Field(min_length=1, max_length=300)


class WhatsAppStepCreate(_StepBase):
    # First-class automated channel (sent via the Baileys sidecar), no subject.
    channel: Literal["whatsapp"]
    subject: None = None
    body: str = Field(min_length=1, max_length=4000)


class ManualTaskStepCreate(_StepBase):
    channel: Literal["call", "sms"]
    subject: str | None = Field(default=None, max_length=250)
    body: str = Field(min_length=1, max_length=4000)


StepCreate = Annotated[
    Union[
        EmailStepCreate,
        LinkedInDmStepCreate,
        LinkedInConnectStepCreate,
        WhatsAppStepCreate,
        ManualTaskStepCreate,
    ],
    Field(discriminator="channel"),
]


class StepOut(BaseModel):
    id: int
    sequence_id: int
    step_order: int
    channel: str
    delay_days: int
    delay_hours: int
    subject: str | None
    body: str
    config: dict
    transitions: list = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


class TransitionsUpdate(BaseModel):
    transitions: list[dict] = Field(default_factory=list)


# ---------- Sequences ----------

SequenceStatus = Literal["draft", "active", "paused", "archived"]


class SequenceCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=4000)
    timezone: str = Field(default="Asia/Kolkata", max_length=64)
    send_window_start: time = time(9, 0)
    send_window_end: time = time(18, 0)
    send_days_mask: int = Field(default=31, ge=1, le=127)
    ai_followups_enabled: bool = False
    track_opens: bool = False
    track_clicks: bool = False


class SequenceUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=4000)
    timezone: str | None = Field(default=None, max_length=64)
    send_window_start: time | None = None
    send_window_end: time | None = None
    send_days_mask: int | None = Field(default=None, ge=1, le=127)
    ai_followups_enabled: bool | None = None
    track_opens: bool | None = None
    track_clicks: bool | None = None


class StatusChange(BaseModel):
    status: SequenceStatus


class AbPromoteRequest(BaseModel):
    label: str = Field(min_length=1, max_length=40)


# Quick channels-based generator (Sequences page → "✨ Generate with AI" panel).
# Distinct from the RAG prompt+grounding generator (SequenceGenerateRequest below).
class QuickGenerateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    description: str = Field(min_length=10, max_length=4000)
    channels: list[Literal["email", "linkedin"]] = Field(min_length=1)
    num_emails: int = Field(default=3, ge=1, le=6)
    timezone: str = Field(default="Asia/Kolkata", max_length=64)
    # Test mode: zero all step delays and open the send window to 24/7 so the
    # whole cadence fires back-to-back regardless of business hours.
    test_mode: bool = Field(default=False)


class SequenceOut(BaseModel):
    id: int
    name: str
    description: str | None
    status: SequenceStatus
    timezone: str
    send_window_start: time
    send_window_end: time
    send_days_mask: int
    ai_followups_enabled: bool
    track_opens: bool = False
    track_clicks: bool = False
    ai_knowledge_id: str | None = None
    step_count: int
    active_enrolments: int
    created_at: datetime
    updated_at: datetime


class SequenceDetail(SequenceOut):
    steps: list[StepOut]


class SequenceGenerateRequest(BaseModel):
    prompt: str = Field(min_length=10, max_length=4000)
    timezone: str = Field(default="Asia/Kolkata", max_length=64)
    document_ids: list[int] = Field(default_factory=list)


class SequenceGenerateResponse(BaseModel):
    sequence: SequenceDetail
    warnings: list[str] = Field(default_factory=list)


class ReorderRequest(BaseModel):
    step_ids: list[int] = Field(min_length=1, max_length=50)


# ---------- Enrolments ----------

EnrolmentStatus = Literal[
    "active", "paused", "done",
    "stopped_reply", "stopped_bounce", "stopped_manual",
    "stopped_archived", "errored",
]


class EnrolmentCreate(BaseModel):
    lead_ids: list[int] = Field(min_length=1, max_length=5000)


class EnrolmentResult(BaseModel):
    enrolled: int
    deduped: int
    skipped_no_identity: int


class EnrolmentOut(BaseModel):
    id: int
    sequence_id: int
    lead_id: int
    contact_snapshot: dict
    status: EnrolmentStatus
    current_step_order: int
    next_send_at: datetime | None
    enrolled_at: datetime
    stopped_at: datetime | None
    stopped_reason: str | None
